File: src/travel_agent/entities/trip_summary_card.py
"""Structured, consumer-facing summary for the active TripRun card."""

from __future__ import annotations

from typing import Literal, Optional

from pydantic import BaseModel, Field


class TripSummaryFact(BaseModel):
    label: str = Field(min_length=1, max_length=24)
    value: str = Field(min_length=1, max_length=80)
    state: Literal["confirmed", "default", "deferred"]


class TripSummaryCard(BaseModel):
    """Stable payload for the compact and expanded research-summary states."""

    headline: str = Field(min_length=1, max_length=80)
    summary: str = Field(min_length=1, max_length=240)
    facts: list[TripSummaryFact] = Field(default_factory=list, max_length=5)
    priorities: list[str] = Field(default_factory=list, max_length=6)
    current_focus: str = Field(min_length=1, max_length=140)
    next_milestone: Optional[str] = Field(default=None, max_length=120)
    compact_line: str = Field(min_length=1, max_length=140)
    requires_user_confirmation: bool = False

File: frontend/src/lib/tripSummaryCard.ts
import type { TripSummaryCard, TripSummaryFact } from '../types/chat';

/**
 * Accept only the deliberately small server contract used by the research
 * summary card.  This keeps the card decoupled from raw workflow state.
 */
export function normalizeTripSummaryCard(
  raw: { summary_card?: Record<string, unknown> } | Record<string, unknown> | null | undefined
): TripSummaryCard | null {
  const card = raw && 'summary_card' in raw ? raw.summary_card : raw;
  if (!card || typeof card !== 'object') return null;

  const value = card as Record<string, unknown>;
  if (
    typeof value.headline !== 'string'
    || typeof value.summary !== 'string'
    || typeof value.current_focus !== 'string'
    || typeof value.compact_line !== 'string'
  ) {
    return null;
  }

  const facts: TripSummaryFact[] = Array.isArray(value.facts)
    ? value.facts.flatMap((fact): TripSummaryFact[] => {
        if (!fact || typeof fact !== 'object') return [];
        const row = fact as Record<string, unknown>;
        if (typeof row.label !== 'string' || typeof row.value !== 'string') return [];
        if (row.state !== 'confirmed' && row.state !== 'default' && row.state !== 'deferred') return [];
        return [{ label: row.label, value: row.value, state: row.state }];
      })
    : [];

  return {
    headline: value.headline,
    summary: value.summary,
    facts,
    priorities: Array.isArray(value.priorities)
      ? value.priorities.filter((item): item is string => typeof item === 'string')
      : [],
    currentFocus: value.current_focus,
    nextMilestone: typeof value.next_milestone === 'string' ? value.next_milestone : null,
    compactLine: value.compact_line,
    requiresUserConfirmation: value.requires_user_confirmation === true,
  };
}
